### PR TITLE
[Core] Add AuraFlow

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Adds Aura Flow from Fal. 

Test code:

```python
from diffusers import AuraFlowPipeline
from diffusers.utils import make_image_grid
import torch


pipeline = AuraFlowPipeline(
	"AuraDiffusion/auradiffusion-v0.1a0", 
	torch_dtype=torch.float16
).to("cuda")

images = pipeline(
    prompt="a cute cat with tiger like looks",
    height=512,
    width=512,
    num_inference_steps=50, 
    num_images_per_prompt=4,
    generator=torch.Generator().manual_seed(666),
    guidance_scale=3.5,
).images
make_image_grid(images, 1, 4).save("demo_hf.png")
```

> [!WARNING]  
> To download the model you must be a member of the AuraDiffusion org. Follow this (internal) Slack [message](https://huggingface.slack.com/archives/C07ATBZ47KQ/p1720062535940439?thread_ts=1720062481.505559&cid=C07ATBZ47KQ). 

Gives:

image

## TODOS

- [ ] Docs
- [x] Tests
- [x] Scheduler (@yiyixuxu would you be able to help out here? I couldn't find a way to use our existing flow matching scheduler in this case)

~Because of the last point above, the noise scheduling code is taken from the original codebase. But I think this PR is still ready for a first review.~

### CODE REVIEW

File: src/diffusers/models/transformers/lavender_transformer_2d.py [hunk 1]
```diff
@@ -0,0 +1,401 @@
+# Copyright 2024 Stability AI, Lavender Flow, The HuggingFace Team. All rights reserved.
```
Comment (sayakpaul) [inline src/diffusers/models/transformers/lavender_transformer_2d.py:1]: New model because it differs from the SD3 one:

* Uses register tokens
* Mixes MMDiT and another kind of simple DiT block (that uses a concatenated `encoder_hidden_states` and `hidden_states`) 
* The final layer norm is different 
* Position embeddings are different (uses learned positional embeddings) 
* The feedforward is different. We only support GeLU and its variants in the feedforward. It uses SwiGLU. 
* No pooled projections.

File: src/diffusers/models/transformers/lavender_transformer_2d.py [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+        # https://arxiv.org/abs/2309.16588
+        # prevents artifacts in the attention maps
+        self.register_tokens = nn.Parameter(torch.randn(1, 8, self.inner_dim) * 0.02)
+
+        self.gradient_checkpointing = False
+
+    def _set_gradient_checkpointing(self, module, value=False):
+        if hasattr(module, "gradient_checkpointing"):
+            module.gradient_checkpointing = value
+
```
Comment (sayakpaul) [inline src/diffusers/models/transformers/lavender_transformer_2d.py:305]: I have deliberately kept additional methods like feedforward chunking, QKV fusion, etc. out of this class because it helps with the initial review.

File: src/diffusers/pipelines/lavender_flow/pipeline_lavender_flow.py [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+        )
+
+        # 6. Denoising loop
+        num_warmup_steps = max(len(timesteps) - num_inference_steps * self.scheduler.order, 0)
+        dt = 1.0 / num_inference_steps
+        dt = (
+            torch.tensor([dt] * effective_batch_size)
+            .to(self.device)
+            .view([effective_batch_size, *([1] * len(latents.shape[1:]))])
+        )
```
Comment (sayakpaul) [inline src/diffusers/pipelines/lavender_flow/pipeline_lavender_flow.py:456]: @yiyixuxu need your help with the scheduler here.

**Review by yiyixuxu — COMMENTED** [review]
very nice!
left some comments mainly on attention processor

File: src/diffusers/pipelines/lavender_flow/pipeline_lavender_flow.py (hunk 3 above)
Comment (yiyixuxu) [reply src/diffusers/pipelines/lavender_flow/pipeline_lavender_flow.py:456]: ok will take a look tomorrow

**Comment by sayakpaul** [comment]
Looking into the test failures 👀

**Comment by sayakpaul** [comment]
Opened a PR to fix the tests https://github.com/huggingface/diffusers/pull/8800.

**Comment by s9anus98a** [comment]
> # What does this PR do?
> 
> Adds Lavender Flow from Fal.
> 
> Test code:
> 
> ```python
> from diffusers import LavenderFlowPipeline
> from diffusers.utils import make_image_grid
> import torch
> 
> 
> pipeline = LavenderFlowPipeline(
> 	"AuraDiffusion/auradiffusion-v0.1a0", 
> 	revision="refs/pr/1",
> 	torch_dtype=torch.float16
> ).to("cuda")
> 
> images = pipeline(
>     prompt="a cute cat with tiger like looks",
>     height=512,
>     width=512,
>     num_inference_steps=50, 
>     num_images_per_prompt=4,
>     generator=torch.Generator().manual_seed(666),
>     guidance_scale=3.5,
> ).images
> make_image_grid(images, 1, 4).save("demo_hf.png")
> ```
> 
> Warning
> 
> To download the model you must be a member of the AuraDiffusion org. Follow this (internal) Slack [message](https://huggingface.slack.com/archives/C07ATBZ47KQ/p1720062535940439?thread_ts=1720062481.505559&cid=C07ATBZ47KQ).
> 
> Gives:
> 
> image
> ## TODOS
> 
>     * [ ]  Docs
> 
>     * [ ]  Tests
> 
>     * [ ]  Scheduler (@yiyixuxu would you be able to help out here? I couldn't find a way to use our existing flow matching scheduler in this case)
> 
> 
> Because of the last point above, the noise scheduling code is taken from the original codebase. But I think this PR is still ready for a first review.

i can't join that slack only for huggingface employee? is the model release for public?

**Comment by sayakpaul** [comment]
@s9anus98a yeah it's not ready for public testing yet.

File: src/diffusers/models/attention_processor.py [hunk 4]
```diff
@@ -117,6 +118,8 @@ def __init__(
         context_pre_only=None,
     ):
         super().__init__()
+        from .normalization import FP32LayerNorm
```
Comment (DN6) [inline src/diffusers/models/attention_processor.py:121]: Why import this here?

File: src/diffusers/models/transformers/auraflow_transformer_2d.py [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+        super().__init__()
+        if hidden_dim is None:
+            hidden_dim = 4 * dim
+
+        n_hidden = int(2 * hidden_dim / 3)
+        n_hidden = find_multiple(n_hidden, 256)
+
+        self.c_fc1 = nn.Linear(dim, n_hidden, bias=False)
+        self.c_fc2 = nn.Linear(dim, n_hidden, bias=False)
+        self.c_proj = nn.Linear(n_hidden, dim, bias=False)
```
Comment (DN6) [inline src/diffusers/models/transformers/auraflow_transformer_2d.py:94]: Could we make the naming here a bit more descriptive? What does "c" mean here?

File: src/diffusers/models/transformers/auraflow_transformer_2d.py [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+# Taken from the original Aura flow inference code.
+# Our feedforward only has GELU but Aura uses SiLU.
+class AuraFlowFeedForward(nn.Module):
+    def __init__(self, dim, hidden_dim=None) -> None:
+        super().__init__()
+        if hidden_dim is None:
+            hidden_dim = 4 * dim
+
+        n_hidden = int(2 * hidden_dim / 3)
+        n_hidden = find_multiple(n_hidden, 256)
```
Comment (DN6) [inline src/diffusers/models/transformers/auraflow_transformer_2d.py:90]: Could we make the variable names a bit more descriptive.

File: src/diffusers/models/transformers/auraflow_transformer_2d.py [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+        dim (`int`): The number of channels in the input and output.
+        num_attention_heads (`int`): The number of heads to use for multi-head attention.
+        attention_head_dim (`int`): The number of channels in each head.
+        is_last (`bool`): Boolean to determine if this is the last block in the model.
+    """
+
+    def __init__(self, dim, num_attention_heads, attention_head_dim, is_last=False):
+        super().__init__()
+        self.is_last = is_last
+        context_norm_type = "ada_norm_continous" if is_last else "ada_norm_zero"
```
Comment (DN6) [inline src/diffusers/models/transformers/auraflow_transformer_2d.py:165]: IMO it is better to have `context_norm_type` in the `__init__` We can set it to the appropriate value when creating the blocks

```python
for i in range(self.config.num_mmdit_layers):
     is_last = i == len(self.config.num_mmdit_layers) - 1
     AuraFlowTransformerBlock(context_norm_type="ada_norm_continuous if is_last else "ada_norm_zero")
```

File: src/diffusers/models/transformers/auraflow_transformer_2d.py [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+        )
+        self.time_step_embed = Timesteps(num_channels=256, downscale_freq_shift=0, scale=1000, flip_sin_to_cos=True)
+        self.time_step_proj = TimestepEmbedding(in_channels=256, time_embed_dim=self.inner_dim)
+
+        self.joint_transformer_blocks = nn.ModuleList(
+            [
+                AuraFlowTransformerBlock(
+                    dim=self.inner_dim,
+                    num_attention_heads=self.config.num_attention_heads,
+                    attention_head_dim=self.config.attention_head_dim,
```
Comment (DN6) [inline src/diffusers/models/transformers/auraflow_transformer_2d.py:275]: Here wouldn't it mean the `context_norm_type` would always be `ada_norm_zero`? We're not passing in information about the whether this is the final block? Do we then need the check defined here
https://github.com/huggingface/diffusers/pull/8796/files#r1668210149

**Review by DN6 — COMMENTED** [review]
It's looking very nice. I think it's good to proceed with the transformer optimizations, chunked ff etc, and can we add some fast tests?

**Comment by bghira** [comment]
this'll need `PeftAdapterMixin` on the transformer module and some analogue to `SD3LoraLoaderMixin`, or, just `SD3LoraLoaderMixin`.

File: src/diffusers/models/attention_processor.py [hunk 9]
```diff
@@ -217,6 +223,17 @@ def __init__(
         if self.context_pre_only is not None and not self.context_pre_only:
             self.to_add_out = nn.Linear(self.inner_dim, self.out_dim, bias=out_bias)
 
+        if qk_norm is not None and added_kv_proj_dim is not None:
+            if qk_norm == "layer_norm":
+                self.norm_added_q = nn.LayerNorm(dim_head, eps=eps)
+                self.norm_added_k = nn.LayerNorm(dim_head, eps=eps)
```
Comment (yiyixuxu) [inline src/diffusers/models/attention_processor.py:229]:
```suggestion
```
let's add this later when we have a model need this configuration, with joint attention + the regular "layer_norm" qk norm

**Comment by bghira** [comment]
i've been testing a fork of this with the LoRA support and it works without any changes to just add the peft adapter to the Transformer2D model and the SD3 LoRA loader mixin to the pipeline.

File: src/diffusers/models/transformers/auraflow_transformer_2d.py (hunk 7 above)
Comment (sayakpaul) [reply src/diffusers/models/transformers/auraflow_transformer_2d.py:165]: `is_last` is enough to determine other block initialization, not just this block. It's something we followed for SD3, too. So, I think this should be okay.

File: src/diffusers/models/transformers/auraflow_transformer_2d.py (hunk 8 above)
Comment (sayakpaul) [reply src/diffusers/models/transformers/auraflow_transformer_2d.py:274]: Resolved.

**Comment by sayakpaul** [comment]
@yiyixuxu @DN6 I have addressed the comments. Here are some extended comments from my end: 

* Run conversion again because of changes needed to address https://github.com/huggingface/diffusers/pull/8796#discussion_r1668193938. Here's the Hub PR: https://huggingface.co/AuraDiffusion/auradiffusion-v0.1a0/discussions/3. 
* I think the transformer optimizations can be tackled later in a separate PR. We saw for SD3 that chunked feedforward doesn't lead to a lot of memory saving. So, I would like to first see how much that gain is and then add accordingly. 
* https://github.com/huggingface/diffusers/actions/runs/9850568702/job/27195993322?pr=8796#step:5:173 (circular import for FP32 layer norm). 

@bghira, I will add LoRA support in an immediate future PR once this PR is merged to keep the reviewing scope concrete and manageable.

**Comment by sayakpaul** [comment]
I have also added fast tests and decided to make the default value of negative prompt to be None instead of "This is watermark, jpeg image white background, web image". I think this is better aligned with our other pipelines. Will include this negative prompt in the docs once I start adding them.

File: src/diffusers/models/transformers/auraflow_transformer_2d.py (hunk 8 above)
Comment (DN6) [reply src/diffusers/models/transformers/auraflow_transformer_2d.py:274]: Sorry, what I mean here is, why even have `if context_norm_type == "ada_norm_continous"` if it isn't used anywhere?

Comment (sayakpaul) [reply src/diffusers/models/transformers/auraflow_transformer_2d.py:264]: Cool. ed3391363 should have addressed it. 

For your next round of review, could I request you to have the comments more on things that might merge blocking (barring documentation)?

File: src/diffusers/models/normalization.py [hunk 10]
```diff
@@ -205,20 +224,28 @@ def __init__(
...
+
         self.silu = nn.SiLU()
         self.linear = nn.Linear(conditioning_embedding_dim, embedding_dim * 2, bias=bias)
         if norm_type == "layer_norm":
             self.norm = LayerNorm(embedding_dim, eps, elementwise_affine, bias)
+        elif norm_type == "fp32_layer_norm":
+            self.norm = FP32LayerNorm(embedding_dim, eps=eps, elementwise_affine=elementwise_affine, bias=bias)
         elif norm_type == "rms_norm":
             self.norm = RMSNorm(embedding_dim, eps, elementwise_affine)
+        elif norm_type == "no_norm":
```
Comment (yiyixuxu) [inline src/diffusers/models/normalization.py:236]: I don't think we should call it `AdaLayerNormContinuous` without the normalization layer here 
I think we can move this out and make a special layer in `auraflow_transformers_2d`

File: src/diffusers/models/transformers/auraflow_transformer_2d.py [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+        hidden_states = hidden_states * (1 + scale_mlp[:, None]) + shift_mlp[:, None]
+        ff_output = self.ff(hidden_states)
+        hidden_states = gate_mlp.unsqueeze(1) * ff_output
+        hidden_states = residual + hidden_states
+
+        return hidden_states
+
+
+@maybe_allow_in_graph
+class AuraFlowTransformerBlock(nn.Module):
```
Comment (yiyixuxu) [inline src/diffusers/models/transformers/auraflow_transformer_2d.py:147]: maybe `AuraFlowJointTransformerBlock`?

File: src/diffusers/models/transformers/auraflow_transformer_2d.py [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+# Taken from the original aura flow inference code.
+def find_multiple(n: int, k: int) -> int:
+    if n % k == 0:
+        return n
+    return n + k - (n % k)
```
Comment (yiyixuxu) [inline src/diffusers/models/transformers/auraflow_transformer_2d.py:39]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py [hunk 13]
```diff
@@ -158,7 +158,12 @@ def scale_noise(
     def _sigma_to_t(self, sigma):
         return sigma * self.config.num_train_timesteps
 
-    def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.device] = None):
+    def set_timesteps(
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py:161]: umm I don't think these changed are introduced in this PR

File: src/diffusers/pipelines/aura_flow/pipeline_aura_flow.py [hunk 14]
```diff
@@ -0,0 +1,505 @@
...
+        max_length = max_sequence_length
+        if prompt_embeds is None:
+            text_inputs = self.tokenizer(
+                prompt,
+                truncation=True,
+                max_length=max_length,
+                padding="max_length",
+                return_tensors="pt",
+            )
+            text_inputs = {k: v.to(device) for k, v in text_inputs.items()}
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/aura_flow/pipeline_aura_flow.py:237]: just curious what else is in `text_inputs` other than the `text_input_ids`?

File: src/diffusers/pipelines/aura_flow/pipeline_aura_flow.py [hunk 15]
```diff
@@ -0,0 +1,505 @@
...
+        generator: Optional[Union[torch.Generator, List[torch.Generator]]] = None,
+        latents: Optional[torch.Tensor] = None,
+        prompt_embeds: Optional[torch.Tensor] = None,
+        prompt_attention_mask: Optional[torch.Tensor] = None,
+        negative_prompt_embeds: Optional[torch.Tensor] = None,
+        negative_prompt_attention_mask: Optional[torch.Tensor] = None,
+        output_type: Optional[str] = "pil",
+        return_dict: bool = True,
+        callback: Optional[Callable[[int, int, torch.Tensor], None]] = None,
+        callback_steps: int = 1,
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/aura_flow/pipeline_aura_flow.py:378]:
```suggestion
```
these are deprecated, I think it is ok to add callback support for now, but if we are adding it, we should use the new callback API https://github.com/huggingface/diffusers/blob/a785992c1d6fcb1ff66f8a0d68d09a0a81b909e8/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L773

**Review by yiyixuxu — COMMENTED** [review]
very nice!!

File: src/diffusers/pipelines/aura_flow/pipeline_aura_flow.py (hunk 14 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/aura_flow/pipeline_aura_flow.py:237]: `attention_mask`

File: src/diffusers/models/transformers/auraflow_transformer_2d.py (hunk 11 above)
Comment (sayakpaul) [reply src/diffusers/models/transformers/auraflow_transformer_2d.py]: Done. Changed `AuraFlowDiTTransformerBlock` to `AuraFlowSingleTransformerBlock` as well for consistency.

File: src/diffusers/models/normalization.py (hunk 10 above)
Comment (sayakpaul) [reply src/diffusers/models/normalization.py]: Done!

File: src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py (hunk 13 above)
Comment (sayakpaul) [reply src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py:161]: @yiyixuxu we merged https://github.com/huggingface/diffusers/issues/8799 into the current PR branch. So, the commits will come here. But they still belong to you, so I guess that is okay?

Comment (yiyixuxu) [reply src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py:161]: you're right, I was confused 😅

**Review by yiyixuxu — APPROVED** [review]
thanks!
